src/pages/why-do-hooks-rely-on-call-order/index.zh-hans.md

```diff
@@ -0,0 +1,601 @@
+
+```js{2,3,7,8}
+// ⚠️ This is NOT the React Hooks API
+const useNameFormInput = createUseFormInput();
+const useSurnameFormInput = createUseFormInput();
+
+function Form() {
+  // ...
+  const name = useNameFormInput();
+  const surname = useNameFormInput();
+  // ...
+}
+```
+
+你还需要非常精确的命名，总是需要考虑“两层”命名 —— 像 `createUseFormInput` 这样的工厂函数和 `useNameFormInput`、`useSurnameFormInput`这样的实例 Hooks。
+
+如果你同时调用两次相同的 custom hook “实例”，你会发生state冲突。事实上，上面的代码就是这种错误 —— 发现了吗？ 它应该为：
+
+```js
+  const name = useNameFormInput();
+  const surname = useSurnameFormInput(); // Not useNameFormInput!
+```
+
+这些问题并非不可克服，但我认为它们会比遵守 [Hooks规则](https://reactjs.org/docs/hooks-rules.html) 的阻力大些。
+
+重要的是，它们打破了复制粘贴的小算盘。在没有封装外层的情况下这种 custom hook 仍然可以使用，但它们只可以被调用一次(这在使用时会产生问题)。不幸的是，当一个API看起来可以正常运行，一旦你意识到在链的某个地方出现了冲突时，就不得不把所有定义好的东西包起来了。
+
+### 缺陷 #6：我们仍然需要一个代码检查工具
+
+还有另外一种使用密钥state来避免冲突的方法，如果你知道，可能会真的很生气，因为我不看好它，抱歉。
+
+这个主意就是每次写 custom hook 时 *组合* 一个密钥，就像这样：
+
+```js{4,5,16,17}
+// ⚠️ This is NOT the React Hooks API
+function Form() {
+  // ...
+  const name = useFormInput('name');
+  const surname = useFormInput('surname');
+  // ...
+  return (
+    <>
+      <input {...name} />
+      <input {...surname} />
+      {/* ... */}
+    </>    
+  )
+}
+
+function useFormInput(formInputKey) {
+  const [value, setValue] = useState('useFormInput(' + formInputKey + ').value');
+  return {
+    value,
+    onChange(e) {
+      setValue(e.target.value);
+    },
+  };
+}
+```
+
+和其他替代提议比，我最不喜欢这个，我觉得它没有什么价值。
+
+一个 Hook 经过多次调用或者与其他 Hook 冲突之后，代码可能 *意外产出* 非唯一或合成无效密钥进行传递。更糟糕的是，如果它是在某些条件下发生的(我们会试图 “修复” 它对吧？)，可能在一段时间后才发生冲突。
+
+我们想提醒大家，记住所有通过密钥来标记的 custom hooks 都很脆弱，它们不仅增加了运行时的工作量(别忘了它们要转成 *密钥* )，而且会渐渐增大 bundle 大小。**但如果说我们非要提醒一个问题，是哪个问题呢**？
+
+如果非要在条件判断里声明 state 和 effects，这种方法可能是有作用的，但按过去经验来说，我发现它令人困惑。事实上，我不记得有人会在条件判断里定义`this.state`或者`componentMount`的。
+
+这段代码到底意味着什么？
+
+```js{3,4}
+// ⚠️ This is NOT the React Hooks API
+function Counter(props) {
+  if (props.isActive) {
+    const [count, setCount] = useState('count');
+    return (
+      <p onClick={() => setCount(count + 1)}>
+        {count}
+      </p>;
+    );
+  }
+  return null;
+}
+```
+
+当 `props.isActive` 为 `false` 时 `count` 是否被保留？或者由于 `useState('count')` 没有被调用而重置 `count`？
+
+如果条件为保留 state，effect 又会发生什么？
+
+```js{5-8}
+// ⚠️ This is NOT the React Hooks API
+function Counter(props) {
+  if (props.isActive) {
+    const [count, setCount] = useState('count');
+    useEffect(() => {
+      const id = setInterval(() => setCount(c => c + 1), 1000);
+      return () => clearInterval(id);
+    }, []);
+    return (
+      <p onClick={() => setCount(count + 1)}>
+        {count}
+      </p>;
+    );
+  }
+  return null;
+}
+```
+
+无疑它不会在 `props.isActive` 第一次是 `true` *之前* 运行，但一旦变成 `true`，它会停止运行吗？当 `props.isActive` 转变为 `false` 时 interval 会重置吗？如果是这样，effect 与 state(我们说不重置时) 的行为不同令人困惑。如果 effect 继续运行，那么 effect 外层的 `if` 不再控制 effect，这也令人感到困惑，我们不是说我们想要基于条件控制的 effects 吗？  
```

Comment:
```suggestion
无疑它不会在 `props.isActive` 第一次是 `true` *之前*运行，但一旦变成 `true`，它会停止运行吗？当 `props.isActive` 转变为 `false` 时 interval 会重置吗？如果是这样，effect 与 state （我们说不重置时）的行为不同令人困惑。如果 effect 继续运行，那么 effect 外层的 `if` 不再控制 effect，这也令人感到困惑，我们不是说我们想要基于条件控制的 effect 吗？  
```
I suggest using singular for the last `effect` to (1) keep it consistent with other `effects`, (2) make it sound more natural since the concept of singular/plural doesn't matter here in Chinese. But this is up to you!